### PR TITLE
planner: refactor the cost implementation for TiFlash plans

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #33945

Problem Summary: planner: refactor the cost implementation for TiFlash plans

### What is changed and how it works?

planner: refactor the cost implementation for TiFlash plans

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -2201,13 +2203,12 @@ func (p *PhysicalHashAgg) attach2TaskForMpp(tasks ...task) task {
...
 		attachPlan2Task(finalAgg, newMpp)
-		if proj != nil {
-			attachPlan2Task(proj, newMpp)
-		}
 		// TODO: how to set 2-phase cost?
 		newMpp.addCost(p.GetCost(inputRows, false, true))
 		finalAgg.SetCost(newMpp.cost())
 		if proj != nil {
+			attachPlan2Task(proj, newMpp)
+			newMpp.addCost(proj.GetCost(newMpp.count()))
```
Comment (qw4990) [inline planner/core/task.go:2211]: The Projection's cost is not accumulated before.

File: planner/core/task.go [hunk 2]
```diff
@@ -1436,6 +1436,8 @@ func (p *PhysicalTopN) attach2Task(tasks ...task) task {
 		copTask.addCost(pushedDownTopN.GetCost(inputCount, false))
 	} else if mppTask, ok := t.(*mppTask); ok && needPushDown && p.canPushDown(kv.TiFlash) {
 		pushedDownTopN := p.getPushedDownTopN(mppTask.p)
+		mppTask.addCost(pushedDownTopN.GetCost(mppTask.p.StatsCount(), false))
```
Comment (qw4990) [inline planner/core/task.go:1439]: The TopN's cost is not accumulated before.

File: planner/core/task.go [hunk 3]
```diff
@@ -2305,7 +2306,7 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
 			attachPlan2Task(p, t)
 		}
 	} else if _, ok := t.(*mppTask); ok {
-		return p.attach2TaskForMpp(tasks...)
+		return final.attach2TaskForMpp(tasks...)
```
Comment (qw4990) [inline planner/core/task.go:2309]: This can make some plan changes in uint tests.

File: planner/core/testdata/integration_suite_out.json [hunk 4]
```diff
@@ -3740,19 +3740,18 @@
           "TableReader 2.00 root  data:ExchangeSender",
           "└─ExchangeSender 2.00 mpp[tiflash]  ExchangeType: PassThrough",
           "  └─Projection 2.00 mpp[tiflash]  Column#9, test.t2.v1, test.t2.v2",
-          "    └─HashAgg 2.00 mpp[tiflash]  group by:test.t2.v1, test.t2.v2, funcs:sum(Column#22)->Column#9, funcs:firstrow(test.t2.v1)->test.t2.v1, funcs:firstrow(test.t2.v2)->test.t2.v2",
+          "    └─HashAgg 2.00 mpp[tiflash]  group by:test.t2.v1, test.t2.v2, funcs:count(1)->Column#9, funcs:firstrow(test.t2.v1)->test.t2.v1, funcs:firstrow(test.t2.v2)->test.t2.v2",
           "      └─ExchangeReceiver 2.00 mpp[tiflash]  ",
           "        └─ExchangeSender 2.00 mpp[tiflash]  ExchangeType: HashPartition, Hash Cols: [name: test.t2.v1, collate: binary], [name: test.t2.v2, collate: binary]",
-          "          └─HashAgg 2.00 mpp[tiflash]  group by:test.t2.v1, test.t2.v2, funcs:count(1)->Column#22",
```
Comment (qw4990) [inline planner/core/testdata/integration_suite_out.json:3746]: This change is caused by `PhysicalHashAgg.attach2Task`.

**Review by Reminiscent — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**